evap/evaluation/management/commands/anonymize.py

```diff
@@ -111,22 +129,100 @@ def replace_usernames_emails(random_usernames):
+            user.save()            
+
+
+    # Comprehensively anonymizes courses.
+    def anonymize_courses(self):
+        all_degrees = Degree.objects.all()
+        all_course_types = CourseType.objects.all()
+
+        # Handling courses of each semester separately in order to prevent
+        # conflicts with archived courses.
         for semester in Semester.objects.all():
-            shuffled_courses = list(semester.courses.all())
-            random.shuffle(shuffled_courses)
-
-            for i, course in enumerate(semester.courses.all()):
-                course.degrees.set(shuffled_courses[i].degrees.all())
-                course.semester = shuffled_courses[i].semester
-                course.name_de = shuffled_courses[i].name_de + " "  # add a space to avoid name collisions
-                course.name_en = shuffled_courses[i].name_en + " "
+            courses = semester.courses.all()
+            public_courses = [course for course in courses if not course.is_private]
+            private_courses = [course for course in courses if course.is_private]
+
+            num_public, num_private = len(public_courses), len(private_courses)
+            self.stdout.write(f"Anonymizing courses of semester {semester} ("
+                + ( f"{num_public} public" if num_private == 0 else 
+                    f"{num_private} private" if num_public == 0 else
+                    f"{num_public} public, {num_private} private, {len(courses)} in total" )
+                + ")...")
+
+            # Randomize the degrees.
+            # Either, this is a Bachelor course or a Master course or a
+            # Bachelor/Master course or something else
+            self.stdout.write("Randomizing degrees...")
+            for course in courses:
+                # Build a set of degrees and add some degrees to it (at least
+                # one).
+                new_degrees = { random.choice(all_degrees) }
+                for __ in range(len(all_degrees)):
+                    if random.random() < 0.5:
+                        new_degrees.add(random.choice(all_degrees))
+                course.degrees.set(new_degrees)
+                course.save()
+
+            # Randomize course types.
+            self.stdout.write("Randomizing course types...")
+            for course in courses:
+                course.type = random.choice(all_course_types)
+                course.save()
+
+            # Assign public courses' names to private ones as their names may
+            # be confidential.
+            # In order to ensure the same course name doesn't appear twice, the
+            # random courses get shuffled and their names then get applied to
+            # the private ones in order. If there are more private courses than
+            # public ones, the remaining private courses just get some default
+            # names.
+            self.stdout.write("Replacing private course names...")
+            random.shuffle(public_courses)
```

Comment:
we need to apply the shuffling to the database to break the link between course names and the results.